dronekit/__init__.py

```diff
@@ -2397,6 +2397,125 @@ def reboot(self):
 
         self.send_mavlink(reboot_msg)
 
+    def calibrate_gyro(self):
+        """Request gyroscope calibration."""
+
+        calibration_command = self.message_factory.command_long_encode(
+            0, 0,  # target_system, target_component
+            mavutil.mavlink.MAV_CMD_PREFLIGHT_CALIBRATION,  # command
+            0,  # confirmation
+            1,  # param 1, 1: gyro calibration, 3: gyro temperature calibration
+            0,  # param 2, 1: magnetometer calibration
+            0,  # param 3, 1: ground pressure calibration
+            0,  # param 4, 1: radio RC calibration, 2: RC trim calibration
+            0,  # param 5, 1: accelerometer calibration, 2: board level calibration, 3: accelerometer temperature calibration, 4: simple accelerometer calibration
+            0,  # param 6, 2: airspeed calibration
+            0,  # param 7, 1: ESC calibration, 3: barometer temperature calibration
+        )
+        self.send_mavlink(calibration_command)
+
+    def calibrate_magnetometer(self):
+        """Request magnetometer calibration."""
+
+        # APM requires the MAV_CMD_DO_START_MAG_CAL command, only present in the APM MAVLink dialect
+        if self._autopilot_type == mavutil.mavlink.MAV_AUTOPILOT_ARDUPILOTMEGA:
+            calibration_command = self.message_factory.command_long_encode(
+                0, 0,  # target_system, target_component
+                mavutil.mavlink.MAV_CMD_DO_START_MAG_CAL,  # command
+                0,  # confirmation
+                0,  # param 1, uint8_t bitmask of magnetometers (0 means all).
+                1,  # param 2, Automatically retry on failure (0=no retry, 1=retry).
+                1,  # param 3, Save without user input (0=require input, 1=autosave).
+                0,  # param 4, Delay (seconds).
+                1,  # param 5, Autoreboot (0=user reboot, 1=autoreboot).
```

Comment:
I think I'd like to see a parameter for this one.

We'd be looking for consistency across vehicles here, I think; ArduPilot should be left in the same state as the PX4 firmware; either requiring a reboot or not, depending on whetherPX4 does or not after a mag cal.

Reply:
PX4 does not reboot automatically after a mag calibration.
Some people say that the reboot isn't sufficient and recommend a full power cycle, however: https://github.com/PX4/Firmware/issues/9030

Reply:
Ok, I disabled auto-reboot for mag calibration on ArduPilot, so that the behaviour is consistent with PX4.
It is always possible to reboot using the `Vehicle.reboot()` method.